### ISSUE TITLE
Tab navigation is extremely slow when the application is run.

### ISSUE THREAD

**Issue by mahiraltinkaya** [react-navigation/react-navigation#7939]
Hello there;

I used to use Stack and Tab navigation before. I also use Redux and as it is known, Redux is injected into the application via props. Stack navigation was extremely slow so I changed all navigation to tab navigation. My current problem is when the application first runs, the first transition between pages is still very slow. However, the page I entered once runs extremely fast afterwards. Why can the problem be caused? (Debug and release mode tested.)

**Comment by ralph-dev** [comment]
https://reactnavigation.org/docs/bottom-tab-navigator#lazy
Try Lazy loading, maybe yours Tabs are loading way too much data on first load.

**Comment by mahiraltinkaya** [comment]
> https://reactnavigation.org/docs/bottom-tab-navigator#lazy
> Try Lazy loading, maybe yours Tabs are loading way too much data on first load.

Thanks i tried it. Faster than before. But still stack navigation slowly with programatically navigation. For that i've deleted all stact navigation configration.

**Comment by ralph-dev** [comment]
One other optimization you could do is reduce the amount of nested navigators you have as well. More Nesting = Slower Preformance on lower end devices.

**Comment by cklinx** [comment]
Have same problem.

**Comment by mahiraltinkaya** [comment]
> Have same problem.


You can try not to use lazy loading on navigator. Really better. 
But programmatically it doesn't matter much.

**Comment by satya164** [comment]
Closing since you didn't provide any repro or didn't fill the issue template. We need a minimal repro to be able to debug issues. Feel free to open a new issue if you have a repro.

**Comment by jeanlambert17** [comment]
Same problem here... It gets really slow even tough the nested navigators only have one or maybe 2 screens

**Comment by shurshilov** [comment]
i have exactly the same problem. In theory, the screen should be displayed with data from redux-persist instantly. I think there is some connection between redux or redux-persist with the slow navigation. Moreover, it is the first transition that slows down, then it works quickly. I use all the latest versions of all libraries for the current date.

**Comment by ralph-dev** [comment]
> i have exactly the same problem. In theory, the screen should be displayed with data from redux-persist instantly. I think there is some connection between redux or redux-persist with the slow navigation. Moreover, it is the first transition that slows down, then it works quickly. I use all the latest versions of all libraries for the current date.
> 
> If i change screen to just Text work quickly, data dont request from API only redux persist
> 
> ```
> 
>   render() {
> 
>     const { search } = this.state;
> 
> //work quickly
> 
>     return (<><Text> test </Text></>)
> 
>     return (
> 
>         <ScrollView style={{ backgroundColor: "white", flex: 1 }}>
> 
>           <SearchBar
> 
>             placeholder="Search contact..."
> 
>             onChangeText={this.updateSearch}
> 
>             value={search}
> 
>           />
> 
> 
> 
>           <View>
> 
>             <Button
> 
>                 title="Update Contact"
> 
>                 onPress={() => this.props.getResPartners()}
> 
>             />
> 
> 
> 
>             {this.props.res_partners.res_partners_loading ? 
> 
>               ( <Text style={styles.text}>
> 
>                   Update contacts ...
> 
>                </Text>
> 
>               ) : ( 
> 
>                 <Text style={styles.text}>
> 
>                     {this.props.res_partners.res_partners_error ? 'Update contacts error' : 'Update contacts success'}
> 
>                 </Text>
> 
>               )
> 
>             }
> 
>                 <Text style={styles.text}>
> 
>                     {this.props.auth.firebase_push_notification ? 'Firebase push enable' : 'Longpolling push enable'}
> 
>                 </Text>
> 
> 
> 
> 
> 
>             {
> 
>               this.props.res_partners.res_partners ?
> 
>               this.props.res_partners.res_partners.map((l, i) => (
> 
>                 <ListItem key={i} bottomDivider>
> 
>                   <Avatar source={{uri: this.props.auth.odoo_api.complete_url+'/web/image/res.partner/'+l.id.toString()+'/image?width=128'}} />
> 
>                   <ListItem.Content>
> 
>                     <ListItem.Title>{l.name}</ListItem.Title>
> 
>                     <ListItem.Subtitle>{l.phone}</ListItem.Subtitle>
> 
>                     <ListItem.Subtitle>{l.function}</ListItem.Subtitle>
> 
>                   </ListItem.Content>
> 
>                    <ListItem.Chevron/>
> 
>                 </ListItem>
> 
>               )): (<Text>Contacts not found</Text>)
> 
>             }
> 
> 
> 
> 
> 
>           </View>
> 
>         
> 
>       </ScrollView>
> 
>     );
> 
>   }
> 
> ```

Can you try removing the image tags from the code and seeing if the tab navigation loads significantly faster? Maybe it's spending time compressing a large image or something like that? Just a thought.

**Comment by shurshilov** [comment]
> > i have exactly the same problem. In theory, the screen should be displayed with data from redux-persist instantly. I think there is some connection between redux or redux-persist with the slow navigation. Moreover, it is the first transition that slows down, then it works quickly. I use all the latest versions of all libraries for the current date.
> > If i change screen to just Text work quickly, data dont request from API only redux persist
> > ```
> > 
> >   render() {
> > 
> >     const { search } = this.state;
> > 
> > //work quickly
> > 
> >     return (<><Text> test </Text></>)
> > 
> >     return (
> > 
> >         <ScrollView style={{ backgroundColor: "white", flex: 1 }}>
> > 
> >           <SearchBar
> > 
> >             placeholder="Search contact..."
> > 
> >             onChangeText={this.updateSearch}
> > 
> >             value={search}
> > 
> >           />
> > 
> > 
> > 
> >           <View>
> > 
> >             <Button
> > 
> >                 title="Update Contact"
> > 
> >                 onPress={() => this.props.getResPartners()}
> > 
> >             />
> > 
> > 
> > 
> >             {this.props.res_partners.res_partners_loading ? 
> > 
> >               ( <Text style={styles.text}>
> > 
> >                   Update contacts ...
> > 
> >                </Text>
> > 
> >               ) : ( 
> > 
> >                 <Text style={styles.text}>
> > 
> >                     {this.props.res_partners.res_partners_error ? 'Update contacts error' : 'Update contacts success'}
> > 
> >                 </Text>
> > 
> >               )
> > 
> >             }
> > 
> >                 <Text style={styles.text}>
> > 
> >                     {this.props.auth.firebase_push_notification ? 'Firebase push enable' : 'Longpolling push enable'}
> > 
> >                 </Text>
> > 
> > 
> > 
> > 
> > 
> >             {
> > 
> >               this.props.res_partners.res_partners ?
> > 
> >               this.props.res_partners.res_partners.map((l, i) => (
> > 
> >                 <ListItem key={i} bottomDivider>
> > 
> >                   <Avatar source={{uri: this.props.auth.odoo_api.complete_url+'/web/image/res.partner/'+l.id.toString()+'/image?width=128'}} />
> > 
> >                   <ListItem.Content>
> > 
> >                     <ListItem.Title>{l.name}</ListItem.Title>
> > 
> >                     <ListItem.Subtitle>{l.phone}</ListItem.Subtitle>
> > 
> >                     <ListItem.Subtitle>{l.function}</ListItem.Subtitle>
> > 
> >                   </ListItem.Content>
> > 
> >                    <ListItem.Chevron/>
> > 
> >                 </ListItem>
> > 
> >               )): (<Text>Contacts not found</Text>)
> > 
> >             }
> > 
> > 
> > 
> > 
> > 
> >           </View>
> > 
> >         
> > 
> >       </ScrollView>
> > 
> >     );
> > 
> >   }
> > ```
> 
> Can you try removing the image tags from the code and seeing if the tab navigation loads significantly faster? Maybe it's spending time compressing a large image or something like that? Just a thought.

it work slowly on any tab first click, withouts images too) but when I delete avatar work faster but first click slowly...

**Comment by ralph-dev** [comment]
lazyPlaceholder
Can you use this prop in the docs, and let me know if an activity indicator appears and for how long for you?

**Comment by shurshilov** [comment]
> lazyPlaceholder
> Can you use this prop in the docs, and let me know if an activity indicator appears and for how long for you?

I create 
`const lazyPlaceholder = () => (
  <View     style={{flex: 1, justifyContent: 'center'}}>
    <ActivityIndicator
      size="large"
    />
  </View>
);`
and add as <Tab.Navigator initialRouteName="ProjectTaskStack" lazy={true} lazyPlaceholder={lazyPlaceholder}
but i dont see indicator

**Comment by shurshilov** [comment]
My app start may be help
`// In App.js in a new project
//import 'react-native-gesture-handler';
import React from 'react';
import AuthNavigation from '../components/router';
import { NavigationContainer } from '@react-navigation/native';
import { Provider as ReduxProvider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga'

import combineReducer from '../reducers/index';
import rootSaga from '../sagas'

import AsyncStorage from '@react-native-community/async-storage';
import { persistStore, persistReducer } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react';


import { enableScreens } from 'react-native-screens';
//import { createNativeStackNavigator } from 'react-native-screens/native-stack';
enableScreens();
//const Stack = createNativeStackNavigator();

const persistConfig = {

    key: 'persistedReducer',
 
    storage: AsyncStorage
}
const persistedReducer = persistReducer(persistConfig, combineReducer)

const sagaMiddleware = createSagaMiddleware();
const store = createStore(persistedReducer, applyMiddleware(sagaMiddleware));
let persistor = persistStore(store)//, {}, () => {
      //console.log('PERSIST CALLBACK')
      //console.log('state', store.getState());
      //store.dispatch({type: "API_INIT"});
      //apiInit();
    //})

sagaMiddleware.run(rootSaga);


export default class App extends React.Component {
  render() {
    return (
        <ReduxProvider store={store}>
            <PersistGate loading={null} persistor={persistor}>
                <NavigationContainer>
                    <AuthNavigation/>
                </NavigationContainer>
            </PersistGate>
        </ReduxProvider>
    );
  }
}
`

**Comment by ralph-dev** [comment]
I'm not really sure, I would imagine it's becuase the navigation service takes a few seconds to load, and becuase it's the first thing your app looks at that could be the issue. The only suggestions I have is to have an app start loading indicator possibly or force pushing these events
App Load -> Go to tab -> go to non tab then when they click on the tab it'll be loaded

**Comment by theambidextrous** [comment]
> > lazyPlaceholder
> > Can you use this prop in the docs, and let me know if an activity indicator appears and for how long for you?
> 
> I create
> `const lazyPlaceholder = () => ( <View style={{flex: 1, justifyContent: 'center'}}> <ActivityIndicator size="large" /> </View> );`
> and add as <Tab.Navigator initialRouteName="ProjectTaskStack" lazy={true} lazyPlaceholder={lazyPlaceholder}
> but i dont see indicator

Assuming that "It i never too late" is true... see below:

<ActivityIndicator/> Component in android requires that you assign a value to its color prop

**Comment by jp928** [comment]
Same issue here. slow on android.

**Comment by MrAlexWeber** [comment]
I'd like to suggest that this may be related to https://github.com/react-navigation/react-navigation/issues/8269 - navigation appears to be blocked by XHR / fetch / API requests.

**Comment by 10000multiplier** [comment]
@mralexweber any solution for this?

**Comment by shurshilov** [comment]
I think its problem only on simulator and i just work with it(((

**Comment by saeedtkh** [comment]
any updates?

**Comment by kastriotkastrati** [comment]
Hi, so I think the problem is that react navigation waits for the component to compute instead of first running the animation. This means that if you have a component that has expensive computations, the transition between tabs is rather slow. I haven't checked the source code to see whether this is the problem, but with this assumption in mind I create a higher order component that solved my problems.

Here's the code if it helps anyone:
```
import LoadingComponent from '@components/Global/LoadingComponent';
import { useEffect, useState } from 'react';
import { InteractionManager } from 'react-native';

// waits for animations to finish before launching component

const afterAnimations = (Component) => ({ ...props }) => {

    const [animationsDone, setAnimationsDone] = useState(false);

    useEffect(() => {
      setAnimationsDone(false);
      InteractionManager.runAfterInteractions(() => {
        setAnimationsDone(true);
      });
    }, []);

    if (!animationsDone) return <LoadingComponent />;

    return <Component {...props} />;
};

export default afterAnimations;
```

**Comment by kastriotkastrati** [comment]
Hi, so I think the problem is that react-navigation waits for the component to compute instead of first running the animation. This means that if you have a component that has expensive computations, the transition between tabs is rather slow. I haven't checked the source code to see whether this is the problem, but with this assumption in mind, I created a higher-order component that solved my problems.

Here's the code if it helps anyone:
```
import LoadingComponent from '@components/Global/LoadingComponent';
import { useEffect, useState } from 'react';
import { InteractionManager } from 'react-native';

// waits for animations to finish before launching component

const afterAnimations = (Component) => ({ ...props }) => {

    const [animationsDone, setAnimationsDone] = useState(false);

    useEffect(() => {
      setAnimationsDone(false);
      InteractionManager.runAfterInteractions(() => {
        setAnimationsDone(true);
      });
    }, []);

    if (!animationsDone) return <LoadingComponent />;

    return <Component {...props} />;
};

export default afterAnimations;
```

**Comment by thangpq** [comment]
I tried change setting on devide (clear cache when restart) => Then restart device
Then everything veryyyyy fasssttt.

Hope this help you.

**Comment by hmdrx** [comment]
> any updates?

did you find any solution?

**Comment by ogmzh** [comment]
> 

how do you clear cache when restart?

**Comment by ogmzh** [comment]
> I tried change setting on devide (clear cache when restart) => Then restart device Then everything veryyyyy fasssttt.
> 
> Hope this help you.
how do you do this?

**Comment by rxeal** [comment]
I found a way to remove the initial delay when switching tabs! Setting `lazy: false` in `screenOptions` ensures all tabs are preloaded, so there's no lag on the first press.  

Here’s the working solution:  

```tsx
<Tabs
  screenOptions={{
    lazy: false,
  }}
>
```

If setting lazy={false} is too heavy, you can pre-render screens in the background while showing a placeholder.

```tsx
<Tabs
  screenOptions={{
    lazy: true,
    renderLazyPlaceholder: () => <LoadingScreen />,
  }}
>
```

**Comment by Hiba-Bazerbashi** [comment]
@rxeal  what about bottom tabs how can i use similar to this  renderLazyPlaceholder: () => <LoadingScreen />, to enhance  startup time
